File: backend/app/services/timeframe.py
"""Timeframe definitions and per-timeframe configuration (spec section 2)."""

from dataclasses import dataclass
from enum import Enum


class Timeframe(str, Enum):
    M1 = "1m"
    M3 = "3m"
    M5 = "5m"
    M10 = "10m"
    M15 = "15m"
    M30 = "30m"
    H1 = "60m"
    D1 = "1d"


@dataclass(frozen=True)
class TimeframeConfig:
    timeframe: Timeframe
    tr_code: str  # "t8452" | "t8451"
    ncnt: int  # minute unit for t8452 (0 for daily)
    qrycnt: int
    trend_lookback: int
    hold_bars: tuple[int, ...] = (5, 10, 25)
    label: str = ""


TIMEFRAME_CONFIGS: dict[Timeframe, TimeframeConfig] = {
    Timeframe.M1: TimeframeConfig(Timeframe.M1, "t8452", 1, 500, 20, (5, 10, 25), "1분"),
    Timeframe.M3: TimeframeConfig(Timeframe.M3, "t8452", 3, 500, 15, (5, 10, 25), "3분"),
    Timeframe.M5: TimeframeConfig(Timeframe.M5, "t8452", 5, 300, 12, (5, 10, 25), "5분"),
    Timeframe.M10: TimeframeConfig(Timeframe.M10, "t8452", 10, 200, 10, (5, 10, 25), "10분"),
    Timeframe.M15: TimeframeConfig(Timeframe.M15, "t8452", 15, 200, 10, (5, 10, 25), "15분"),
    Timeframe.M30: TimeframeConfig(Timeframe.M30, "t8452", 30, 100, 8, (5, 10, 25), "30분"),
    Timeframe.H1: TimeframeConfig(Timeframe.H1, "t8452", 60, 100, 8, (5, 10, 25), "60분"),
    Timeframe.D1: TimeframeConfig(Timeframe.D1, "t8451", 0, 60, 5, (5, 10, 25), "일봉"),
}

# Upper timeframes for MTF confluence (section 9-3).
MTF_GROUPS: dict[Timeframe, list[Timeframe]] = {
    Timeframe.M5: [Timeframe.M15, Timeframe.H1],
    Timeframe.M15: [Timeframe.H1, Timeframe.D1],
    Timeframe.M30: [Timeframe.H1, Timeframe.D1],
    Timeframe.H1: [Timeframe.D1],
    Timeframe.D1: [],
}


def config_for(tf: Timeframe) -> TimeframeConfig:
    return TIMEFRAME_CONFIGS[tf]

File: backend/src/timeframe.rs
//! Timeframe definitions + per-timeframe configuration (spec section 2).

use serde::{Deserialize, Serialize};

/// Supported candle timeframes (1m … 1d).
#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug, Serialize, Deserialize)]
pub enum Timeframe {
    M1,
    M3,
    M5,
    M10,
    M15,
    M30,
    H1,
    D1,
}

impl Timeframe {
    /// Wire value used by the API / frontend (e.g. `"5m"`, `"1d"`).
    pub fn as_str(self) -> &'static str {
        match self {
            Timeframe::M1 => "1m",
            Timeframe::M3 => "3m",
            Timeframe::M5 => "5m",
            Timeframe::M10 => "10m",
            Timeframe::M15 => "15m",
            Timeframe::M30 => "30m",
            Timeframe::H1 => "60m",
            Timeframe::D1 => "1d",
        }
    }

    /// Parse a wire value; `None` for unknown strings (→ HTTP 400 upstream).
    pub fn parse(v: &str) -> Option<Self> {
        Some(match v {
            "1m" => Timeframe::M1,
            "3m" => Timeframe::M3,
            "5m" => Timeframe::M5,
            "10m" => Timeframe::M10,
            "15m" => Timeframe::M15,
            "30m" => Timeframe::M30,
            "60m" => Timeframe::H1,
            "1d" => Timeframe::D1,
            _ => return None,
        })
    }

    /// Per-timeframe fetch + trend config. (Daily uses t8451; the rest t8452.)
    pub fn config(self) -> TimeframeConfig {
        // (ncnt, qrycnt, trend_lookback)
        let (ncnt, qrycnt, lookback) = match self {
            Timeframe::M1 => (1, 500, 20),
            Timeframe::M3 => (3, 500, 15),
            Timeframe::M5 => (5, 300, 12),
            Timeframe::M10 => (10, 200, 10),
            Timeframe::M15 => (15, 200, 10),
            Timeframe::M30 => (30, 100, 8),
            Timeframe::H1 => (60, 100, 8),
            Timeframe::D1 => (0, 60, 5),
        };
        TimeframeConfig { ncnt, qrycnt, trend_lookback: lookback }
    }

    /// Higher timeframes used for MTF confluence (spec section 9-3).
    pub fn mtf_group(self) -> Vec<Timeframe> {
        use Timeframe::*;
        match self {
            M1 => vec![M5, M15],
            M3 => vec![M15, H1],
            M5 => vec![M15, H1],
            M10 => vec![M30, H1],
            M15 => vec![H1, D1],
            M30 => vec![H1, D1],
            H1 => vec![D1],
            D1 => vec![],
        }
    }
}

/// Static configuration for a single timeframe.
pub struct TimeframeConfig {
    pub ncnt: i32,
    pub qrycnt: i32,
    pub trend_lookback: usize,
}
